<a href="https://colab.research.google.com/github/songqsh/foo1/blob/master/src/21d2210midterm_a_p2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Simplex method

We are going to illustrate simplex tableau method for the dual of the diet problem:
$$\max z = 500 x_1 + 6x_2 + 10 + 8 x_4$$
s.t.
$$\begin{array}{ll}
400 x_1 + 3x_2+2x_3+2x_4 & \le 50\\
200 x_1 + 2x_2 + 2x_3 + 4x_4 & \le 20\\
150x_1 \quad \quad \quad + 4x_3 + x_4 & \le 30\\
500x_1 \quad \quad \quad + 4x_3 + 5x_4 & \le 80 \\
x_1, x_2, x_3, x_4\ge 0.
\end{array}
$$

In [4]:
# import package
import numpy as np
import numpy.linalg as la
float_formatter = "{:.2f}".format
np.set_printoptions(formatter={'float_kind':float_formatter})
import pandas as pd


## Pivotizing

Pivotizing with $(i,j)$-entry means converting the matrix by multiple EROS so that the resulting matrix has all zeros in $j$-column, but one in $(i,j)$-entry. This is the key step in updating basic feasible solution.

In [2]:
#pivotize with (i,j)
def pivot(A, i, j):
  A[i] = A[i]/A[i,j] #scale to get one in (i,j)
  n_rows, _ = A.shape
  for k in range(n_rows):
    if k==i:
      continue # skip i-row
    A[k] = A[k] - A[i]*A[k,j] # replacement to get zero
    

## Initial tableau

In [6]:
A = np.array([[1, -1, -2., -1., 0., 0, 0],
              [0, 1, 1, 0, 1, 0, 6],
              [0, 0, 1, 1, 0, 1, 8]])
n_row, n_col = A.shape

print('tableau has ' + str(n_row) + ' rows and ' + str(n_col) + ' columns')
print(pd.DataFrame(A))

tableau has 3 rows and 7 columns
     0    1    2    3    4    5    6
0  1.0 -1.0 -2.0 -1.0  0.0  0.0  0.0
1  0.0  1.0  1.0  0.0  1.0  0.0  6.0
2  0.0  0.0  1.0  1.0  0.0  1.0  8.0


In [7]:
bv = [4, 5] #set bv
print('bvs are: \n')
print(bv)
#optimal test and pivot column
r0 = list(A[0])
print(r0)
min_value = min(r0)
if min_value >= 0:
  print('done')
else:
  print('continue')
  pivot_col = r0.index(min_value)
  print('pivot column is ' + str(pivot_col))


bvs are: 

[4, 5]
[1.0, -1.0, -2.0, -1.0, 0.0, 0.0, 0.0]
continue
pivot column is 2


In [8]:
#minimum ratio test and pivot row
min_ratio = 10000.0 #big number
pivot_row = 100 #big integer
for i in range(1,n_row):
  if A[i,pivot_col] >0:
    now_ratio = A[i, -1]/A[i, pivot_col]
    if now_ratio < min_ratio:
      min_ratio = now_ratio
      pivot_row = i
print('pivot row is ' + str(pivot_row))
pivot(A, pivot_row,pivot_col)

pivot row is 1


Current tableau

In [9]:
print(pd.DataFrame(A))

     0    1    2    3    4    5     6
0  1.0  1.0  0.0 -1.0  2.0  0.0  12.0
1  0.0  1.0  1.0  0.0  1.0  0.0   6.0
2  0.0 -1.0  0.0  1.0 -1.0  1.0   2.0


In [10]:
bv = [2, 5] #set bv
print('bvs are: \n')
print(bv)
#optimal test and pivot column
r0 = list(A[0])
print(r0)
min_value = min(r0)
if min_value >= 0:
  print('done')
else:
  print('continue')
  pivot_col = r0.index(min_value)
  print('pivot column is ' + str(pivot_col))


bvs are: 

[2, 5]
[1.0, 1.0, 0.0, -1.0, 2.0, 0.0, 12.0]
continue
pivot column is 3


In [11]:

#minimum ratio test and pivot row
min_ratio = 10000.0 #big number
pivot_row = 100 #big integer
for i in range(1,n_row):
  if A[i,pivot_col] >0:
    now_ratio = A[i, -1]/A[i, pivot_col]
    if now_ratio < min_ratio:
      min_ratio = now_ratio
      pivot_row = i
print('pivot row is ' + str(pivot_row))
pivot(A, pivot_row,pivot_col)

pivot row is 2


Current tableau

In [12]:
pd.DataFrame(A)

0    1    2    3    4    5     6
0  1.0  0.0  0.0  0.0  1.0  1.0  14.0
1  0.0  1.0  1.0  0.0  1.0  0.0   6.0
2  0.0 -1.0  0.0  1.0 -1.0  1.0   2.0

In [ ]:
bv = [1, 3, 5, 8] #set bv
print(bv)
#optimal test and pivot column
r0 = list(A[0])
print(r0)
min_value = min(r0)
if min_value >= 0:
  print('done')
else:
  print('continue')
  pivot_col = r0.index(min_value)
  print('pivot column is ' + str(pivot_col))

#minimum ratio test and pivot row
min_ratio = 10000.0 #big number
pivot_row = 100 #big integer
for i in range(1,n_row):
  if A[i,pivot_col] >0:
    now_ratio = A[i, -1]/A[i, pivot_col]
    if now_ratio < min_ratio:
      min_ratio = now_ratio
      pivot_row = i
print('pivot row is ' + str(pivot_row))
pivot(A, pivot_row,pivot_col)

[1, 3, 5, 8]
[1.0, 0.0, -4.0, 0.0, -2.0, 0.0, 1.0, 2.0, 0.0, 80.0]
continue
pivot column is 2
pivot row is 2


Current tableau

In [ ]:
A

array([[1.00, 250.00, 0.00, 0.00, 5.00, 0.00, 3.00, 1.00, 0.00, 90.00],
       [0.00, 137.50, 0.00, 0.00, -3.75, 1.00, -1.50, 0.25, 0.00, 27.50],
       [0.00, 62.50, 1.00, 0.00, 1.75, 0.00, 0.50, -0.25, 0.00, 2.50],
       [0.00, 37.50, 0.00, 1.00, 0.25, 0.00, 0.00, 0.25, 0.00, 7.50],
       [0.00, 350.00, 0.00, 0.00, 4.00, 0.00, 0.00, -1.00, 1.00, 50.00]])

In [ ]:
bv = [2, 3, 5, 8] #set bv
print('bvs are: \n')
print(bv)
#optimal test and pivot column
r0 = list(A[0])
print(r0)
min_value = min(r0)
if min_value >= 0:
  print('done')
else:
  print('continue')
  pivot_col = r0.index(min_value)
  print('pivot column is ' + str(pivot_col))


bvs are: 

[2, 3, 5, 8]
[1.0, 250.0, 0.0, 0.0, 5.0, 0.0, 3.0, 1.0, 0.0, 90.0]
done


## Conclusion

The optimal solution for the dual problem is 
$$x = (0, 2.5, 7.5, 0), z = 90.$$
Indeed, in the original diet problem, the optimal solution is by eating no brownies, 3 oz ice cream, 1 cola, no pineapples, which yields minimum cost 90.